src/java/com/unifina/signalpath/ModuleWithUI.java

```diff
@@ -32,13 +34,38 @@
 
 	private static final Logger log = Logger.getLogger(ModuleWithUI.class);
 
+	private Map<String,Long> previousTimestamps = new HashMap<>();
+	private Map<String,Long> previousSequenceNumbers = new HashMap<>();
+
 	public ModuleWithUI() {
 		super();
 	}
 
+	private long getNextSequenceNumber(String streamId, long timestamp) {
```

Comment:
Can't this functionality (of keeping track of latest MessageRef and sequence number by stream) be abstracted into some utility class
`public class MessageChainUtil implements Serializable { ... }`
and reused here and in `SendToStream`?

Also the chaining should happen by stream-partition I guess?

```
MessageChainUtil msgChainUtil = new MessageChainUtil();

// ...

// Returns the latest known MessageRef (created with .nextMessageRef(...))
MessageRef previousRef = msgChainUtil.getLatestMessageRef(stream, partition)

// Creates and caches a new MessageRef for stream-partition, increments sequence number
MessageRef thisMsgRef = msgChainUtil.nextMessageRef(stream, partition, timestamp)
```

(Also please add unit tests for this util class)